Cannot use `const enum` to define an interface key in `isolatedModules` mode

# Bug Report

I am maintaning a library where I use `const enum` variants to declare the keys of an interface (`interface Foo { [MyConstEnum.Foo]: "bar" }`). The enum and the interface are in the same module.

It works well, but once compiled using `tsc` and included in a user project using `isolatedModules: true`, TypeScript fails with the following error:

>  Error: [...].d.ts: error TS2748: Cannot access ambient const enums when 'isolatedModules' is enabled.

I don't understand why this usage is restricted, since the enum is only used in a type definition, and both are present in the same `.d.ts` file.

### 🔎 Search Terms

Cannot access ambient const enums when the '--isolatedModules' flag is provided.
TS2748
interface

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about isolated modules and const enum

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?isolatedModules=true&ts=5.0.4#code/PTAEEkDtQewVwE6gDYEsBGAaUB3ApqACZ4DGyAhggeaCTJAM4AuoekcAtrqkwBZGoAZoLxVILPADc2LSOQ54G2Pm1DloAIikzQHcgAcNahgxglU5JqkgBzNdG3jQTAJ76CgmEjzkS-R7LyeAB0ALAAUCCgACq8MAwEcgoMalSgkDAs1kxsxITOMKDoBHAJ+eguzrwEaOgIlJWlogzBoAASonjY+ET0AOQS0g4AHvpeLCocYeERdIwS7FwAokNMAHJBoADeEaB7oABiAPJHoAC8oACMmLv7AEIAggBK56AATBEAvhERru6gKxkAFkDK8duF9qAANqA8QbBTBY5HAC6AC50pxiggANy3PYw1bwkKPJ5o0DMBDWGy48KfGkRKKxVApHBeADWKUE1i6RTgLHoJAITHqjH0qGQeHKlQGDBIfWU1VAAANgoRgkwGErQFyJbR6ExyNYUqYFHwqShUGyhbxmaifuFiGRKAQ5sxWIsAYSgm9tnjDidXtc-STXh9afa-gQAAoIGDoCV6KwkWFMEH6H0XcGQgkyIlvREnMmLLE07MpvPBElkilUml0+1RADq1Wg8CQTQQKVdnGtNQw9QQlWsVTwqCQ+ljACtSExsEIR5U5lybIgCNE3HgAMokSn6Fg4Hj8JXMmAUHKEIEwQhwCUMdHCuB4JUKjyn5AwA+2VgIWNIBhxHAUjgfQ7UiMA9gAYXUDIWF8QUTDUDh0FQHRXQWTgWRbUA+hPM9JUva9bz6UBmXdch40laYogATXgWh1HJPACB4NR0BgaRQAAPQbMA1g-bBwCKPA0G0KpSNEX8SJSaw6B-GdsAYGS8FAqI9gAKlwAhCH6WCSHgrt6DdNgMO6Ah6GQRoEhHMdnA3UAAAphxoJVIy1R0KHqKx6AASgZcDQA07RBzNL9SPc518mHFRyU2DgrxvPAeNAB4UhoVkEDZSh4EgQhsHCykvwY7lhmZA0nDQ7g+BHGKFFASRyGQR9UpSaLjK4BI8A4FrCnStlQIifKXUMlhy29FKIOG9Es32JF0VGhR8yRYNnjmr0FsrZ4vgjOzG3ZKkUzTDNfQhfYczhMaGAm+YCxRdFi1EUtTvmvA3nG4aNtJdEa1sOtsSAA)

### 💻 Code

* Using `isolatedModules: true`
* In a `.d.ts` file

```ts
declare const enum EventName {
    FOO = 1,
    BAR = 2
}

type EventMap = {
    [EventName.FOO]: number;
    [EventName.BAR]: string;
};
```

### 🙁 Actual behavior

TS fails with the following error:

> Error: [...].d.ts: error TS2748: Cannot access ambient const enums when 'isolatedModules' is enabled.


### 🙂 Expected behavior

TS succeeds

### Workaround

As a workaround, I declare a fake object using the same keys and values as the `enum`, and then use this to declare the `interface`:

```ts
declare const enum EventName {
    FOO = 1,
    BAR = 2
}

declare const EventNameAsConst: {
  FOO: EventName.FOO
  BAR: EventName.BAR
}

type EventMap = {
    [EventNameAsConst.FOO]: number;
    [EventNameAsConst.BAR]: string;
};
```

This works, but it is a bit cumbersome to have to maintain this intermediary object.

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -37819,7 +37819,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         if (getIsolatedModules(compilerOptions)) {
             Debug.assert(!!(type.symbol.flags & SymbolFlags.ConstEnum));
             const constEnumDeclaration = type.symbol.valueDeclaration as EnumDeclaration;
-            if (constEnumDeclaration.flags & NodeFlags.Ambient) {
+            if (constEnumDeclaration.flags & NodeFlags.Ambient && !(node.flags & NodeFlags.Ambient)) {
                 error(node, Diagnostics.Cannot_access_ambient_const_enums_when_0_is_enabled, isolatedModulesLikeFlagName);
             }
         }

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -37819,7 +37819,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         if (getIsolatedModules(compilerOptions)) {
             Debug.assert(!!(type.symbol.flags & SymbolFlags.ConstEnum));
             const constEnumDeclaration = type.symbol.valueDeclaration as EnumDeclaration;
-            if (constEnumDeclaration.flags & NodeFlags.Ambient && !(node.flags & NodeFlags.Ambient)) {
+            if (constEnumDeclaration.flags & NodeFlags.Ambient && !isValidTypeOnlyAliasUseSite(node)) {
                 error(node, Diagnostics.Cannot_access_ambient_const_enums_when_0_is_enabled, isolatedModulesLikeFlagName);
             }
         }

--- a/tests/baselines/reference/isolatedModulesConstEnum.js
+++ b/tests/baselines/reference/isolatedModulesConstEnum.js
@@ -1,0 +1,21 @@
+//// [tests/cases/compiler/isolatedModulesConstEnum.ts] ////
+
+//// [foo.d.ts]
+declare const enum EventName {
+    FOO = 1,
+    BAR = 2
+}
+
+type E1 = {
+    [EventName.FOO]: number;
+    [EventName.BAR]: string;
+};
+
+//// [bar.ts]
+type E2 = {
+    [EventName.FOO]: number;
+    [EventName.BAR]: string;
+};
+
+
+//// [bar.js]

--- a/tests/baselines/reference/isolatedModulesConstEnum.symbols
+++ b/tests/baselines/reference/isolatedModulesConstEnum.symbols
@@ -9,17 +9,35 @@ declare const enum EventName {
 >BAR : Symbol(EventName.BAR, Decl(foo.d.ts, 1, 12))
 }
 
-type EventMap = {
->EventMap : Symbol(EventMap, Decl(foo.d.ts, 3, 1))
+type E1 = {
+>E1 : Symbol(E1, Decl(foo.d.ts, 3, 1))
 
     [EventName.FOO]: number;
->[EventName.FOO] : Symbol([EventName.FOO], Decl(foo.d.ts, 5, 17))
+>[EventName.FOO] : Symbol([EventName.FOO], Decl(foo.d.ts, 5, 11))
 >EventName.FOO : Symbol(EventName.FOO, Decl(foo.d.ts, 0, 30))
 >EventName : Symbol(EventName, Decl(foo.d.ts, 0, 0))
 >FOO : Symbol(EventName.FOO, Decl(foo.d.ts, 0, 30))
 
     [EventName.BAR]: string;
 >[EventName.BAR] : Symbol([EventName.BAR], Decl(foo.d.ts, 6, 28))
+>EventName.BAR : Symbol(EventName.BAR, Decl(foo.d.ts, 1, 12))
+>EventName : Symbol(EventName, Decl(foo.d.ts, 0, 0))
+>BAR : Symbol(EventName.BAR, Decl(foo.d.ts, 1, 12))
+
+};
+
+=== /bar.ts ===
+type E2 = {
+>E2 : Symbol(E2, Decl(bar.ts, 0, 0))
+
+    [EventName.FOO]: number;
+>[EventName.FOO] : Symbol([EventName.FOO], Decl(bar.ts, 0, 11))
+>EventName.FOO : Symbol(EventName.FOO, Decl(foo.d.ts, 0, 30))
+>EventName : Symbol(EventName, Decl(foo.d.ts, 0, 0))
+>FOO : Symbol(EventName.FOO, Decl(foo.d.ts, 0, 30))
+
+    [EventName.BAR]: string;
+>[EventName.BAR] : Symbol([EventName.BAR], Decl(bar.ts, 1, 28))
 >EventName.BAR : Symbol(EventName.BAR, Decl(foo.d.ts, 1, 12))
 >EventName : Symbol(EventName, Decl(foo.d.ts, 0, 0))
 >BAR : Symbol(EventName.BAR, Decl(foo.d.ts, 1, 12))

--- a/tests/baselines/reference/isolatedModulesConstEnum.symbols
+++ b/tests/baselines/reference/isolatedModulesConstEnum.symbols
@@ -1,0 +1,28 @@
+=== /foo.d.ts ===
+declare const enum EventName {
+>EventName : Symbol(EventName, Decl(foo.d.ts, 0, 0))
+
+    FOO = 1,
+>FOO : Symbol(EventName.FOO, Decl(foo.d.ts, 0, 30))
+
+    BAR = 2
+>BAR : Symbol(EventName.BAR, Decl(foo.d.ts, 1, 12))
+}
+
+type EventMap = {
+>EventMap : Symbol(EventMap, Decl(foo.d.ts, 3, 1))
+
+    [EventName.FOO]: number;
+>[EventName.FOO] : Symbol([EventName.FOO], Decl(foo.d.ts, 5, 17))
+>EventName.FOO : Symbol(EventName.FOO, Decl(foo.d.ts, 0, 30))
+>EventName : Symbol(EventName, Decl(foo.d.ts, 0, 0))
+>FOO : Symbol(EventName.FOO, Decl(foo.d.ts, 0, 30))
+
+    [EventName.BAR]: string;
+>[EventName.BAR] : Symbol([EventName.BAR], Decl(foo.d.ts, 6, 28))
+>EventName.BAR : Symbol(EventName.BAR, Decl(foo.d.ts, 1, 12))
+>EventName : Symbol(EventName, Decl(foo.d.ts, 0, 0))
+>BAR : Symbol(EventName.BAR, Decl(foo.d.ts, 1, 12))
+
+};
+

--- a/tests/baselines/reference/isolatedModulesConstEnum.types
+++ b/tests/baselines/reference/isolatedModulesConstEnum.types
@@ -1,0 +1,30 @@
+=== /foo.d.ts ===
+declare const enum EventName {
+>EventName : EventName
+
+    FOO = 1,
+>FOO : EventName.FOO
+>1 : 1
+
+    BAR = 2
+>BAR : EventName.BAR
+>2 : 2
+}
+
+type EventMap = {
+>EventMap : { 1: number; 2: string; }
+
+    [EventName.FOO]: number;
+>[EventName.FOO] : number
+>EventName.FOO : EventName.FOO
+>EventName : typeof EventName
+>FOO : EventName.FOO
+
+    [EventName.BAR]: string;
+>[EventName.BAR] : string
+>EventName.BAR : EventName.BAR
+>EventName : typeof EventName
+>BAR : EventName.BAR
+
+};
+

--- a/tests/baselines/reference/isolatedModulesConstEnum.types
+++ b/tests/baselines/reference/isolatedModulesConstEnum.types
@@ -11,8 +11,26 @@ declare const enum EventName {
 >2 : 2
 }
 
-type EventMap = {
->EventMap : { 1: number; 2: string; }
+type E1 = {
+>E1 : { 1: number; 2: string; }
+
+    [EventName.FOO]: number;
+>[EventName.FOO] : number
+>EventName.FOO : EventName.FOO
+>EventName : typeof EventName
+>FOO : EventName.FOO
+
+    [EventName.BAR]: string;
+>[EventName.BAR] : string
+>EventName.BAR : EventName.BAR
+>EventName : typeof EventName
+>BAR : EventName.BAR
+
+};
+
+=== /bar.ts ===
+type E2 = {
+>E2 : { 1: number; 2: string; }
 
     [EventName.FOO]: number;
 >[EventName.FOO] : number

--- a/tests/cases/compiler/isolatedModulesConstEnum.ts
+++ b/tests/cases/compiler/isolatedModulesConstEnum.ts
@@ -1,0 +1,11 @@
+// @isolatedModules: true
+// @filename: /foo.d.ts
+declare const enum EventName {
+    FOO = 1,
+    BAR = 2
+}
+
+type EventMap = {
+    [EventName.FOO]: number;
+    [EventName.BAR]: string;
+};

--- a/tests/cases/compiler/isolatedModulesConstEnum.ts
+++ b/tests/cases/compiler/isolatedModulesConstEnum.ts
@@ -5,7 +5,13 @@ declare const enum EventName {
     BAR = 2
 }
 
-type EventMap = {
+type E1 = {
+    [EventName.FOO]: number;
+    [EventName.BAR]: string;
+};
+
+// @filename: /bar.ts
+type E2 = {
     [EventName.FOO]: number;
     [EventName.BAR]: string;
 };

